### PR TITLE
[WIP][KYUUBI-174] define the interface classes of KyuubiSession to allow different implementation of those #120 #174

### PR DESCRIPTION
**Pull request by turboFei**
**Is your feature request related to a problem? Please describe.**

see the description in https://github.com/yaooqinn/kyuubi/issues/120

**Describe the solution you'd like**

I will define the interface classes: IKyuubiSession and IKyuubiOperation.
And  abstract classes: AbstractKyuubiSession and AbstractKyuubiOperation.
And rename KyuubiSession to KyuubiClientSession.

### CODE REVIEW

**Comment by yaooqinn** [comment]
what about separate this pull request to "define the relative parameters" and "abstract kyuubi session and operation"? the don't have any logic in common. AND WHAT is "define the relative parameters" for?

**Comment by hustfeiwang** [comment]
> what about separate this pull request to "define the relative parameters" and "abstract kyuubi session and operation"? the don't have any logic in common. AND WHAT is "define the relative parameters" for?

I will put the content of defining the relative parameters of kyuubi-cluster mode into KYUUBI-175.

**Comment by hustfeiwang** [comment]
Done.

File: kyuubi-server/src/main/scala/yaooqinn/kyuubi/operation/Operation.scala [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  def run(): Unit = {
+    createOperationLog()
+    try {
+      runInternal()
+    } finally {
+      unregisterOperationLog()
+    }
+  }
+
+  protected[this] def cleanupOperationLog(): Unit = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/yaooqinn/kyuubi/operation/Operation.scala]: remove [this]

File: kyuubi-server/src/main/scala/yaooqinn/kyuubi/operation/OperationManager.scala [hunk 2]
```diff
@@ -20,28 +20,27 @@ package yaooqinn.kyuubi.operation
...
+import yaooqinn.kyuubi.session.{KyuubiSession, Session}
 
 private[kyuubi] class OperationManager private(name: String)
   extends AbstractService(name) with Logging {
 
   def this() = this(classOf[OperationManager].getSimpleName)
 
   private[this] lazy val logSchema: StructType = new StructType().add("operation_log", "string")
-  private[this] val handleToOperation = new ConcurrentHashMap[OperationHandle, KyuubiOperation]
+  private[this] val handleToOperation = new ConcurrentHashMap[OperationHandle, Operation]
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/yaooqinn/kyuubi/operation/OperationManager.scala]: remove [this]

File: kyuubi-server/src/main/scala/yaooqinn/kyuubi/operation/OperationManager.scala [hunk 3]
```diff
@@ -20,28 +20,27 @@ package yaooqinn.kyuubi.operation
...
   private[this] val userToOperationLog = new ConcurrentHashMap[String, OperationLog]()
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/yaooqinn/kyuubi/operation/OperationManager.scala]: remove [this] too and other [this]s in this class

**Comment by yaooqinn** [comment]
modify your pr title to remove #174, and add https://github.com/yaooqinn/kyuubi/issues/120 this to the title.

File: kyuubi-server/src/main/scala/org/apache/spark/KyuubiConf.scala [hunk 4]
```diff
@@ -432,6 +432,74 @@ object KyuubiConf {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefault(TimeUnit.SECONDS.toMillis(60L))
 
+  val YARN_KYUUBISERVER_SESSION_MODE: ConfigEntry[String] =
+    KyuubiConfigBuilder("spark.kyuubi.yarn.seesion.mode")
```
[the remark this answers is not in the document]
Comment (hustfeiwang) [reply kyuubi-server/src/main/scala/org/apache/spark/KyuubiConf.scala]: Ok, I have move these content to KYUUBI-175, and i would modify the kyuubiConf  in another PR.

File: kyuubi-server/src/main/scala/org/apache/spark/KyuubiConf.scala [hunk 5]
```diff
@@ -432,6 +432,74 @@ object KyuubiConf {
...
+
+  val YARN_KYUUBIAPPMASTER_USERNAME: ConfigEntry[String] =
+    KyuubiConfigBuilder("spark.kyuubi.yarn.am.user")
+      .doc("The relative user name for the kyuubiAppMaster and it must be specified.")
+      .stringConf
+      .createWithDefault("default")
+
+  val YARN_KYUUBIAPPMASTER_PUBLISH_WAITTIME: ConfigEntry[Long] =
+    KyuubiConfigBuilder("spark.kyuubi.yarn.am.publish.timeout")
+      .doc("Timeout for kyuubiServer to wait kyuubiAppMaster publish service")
```
[the remark this answers is not in the document]
Comment (hustfeiwang) [reply kyuubi-server/src/main/scala/org/apache/spark/KyuubiConf.scala]: a user/session

File: kyuubi-server/src/main/scala/org/apache/spark/KyuubiConf.scala [hunk 6]
```diff
@@ -432,6 +432,74 @@ object KyuubiConf {
...
+      .createWithDefault(false)
+
+  val YARN_KYUUBIAPPMASTER_NAMESPACE: ConfigEntry[String] =
+    KyuubiConfigBuilder("spark.kyuubi.yarn.am.namespace")
+      .doc("The parent node in ZooKeeper used by KyuubiAppMaster when supporting dynamic service" +
+        " discovery and should be different witht that of kyuubiServer.")
+      .stringConf
+      .createWithDefault("kyuubiappmaster")
+
+  val YARN_KYUUBIAPPMASTER_USERNAME: ConfigEntry[String] =
```
[the remark this answers is not in the document]
Comment (hustfeiwang) [reply kyuubi-server/src/main/scala/org/apache/spark/KyuubiConf.scala]: This is used to make a kyuubiAppMaster instance: kyuubiAppMasterNameSpace/username/URIPATH.

**Comment by hustfeiwang** [comment]
I have reconstruct the code and I will reply these  comments  about kyuubiConf in another PR.

**Pull request closed without merging**